### PR TITLE
feat/parsec: dump and reload KeyGen

### PR DESCRIPTION
**Pull request by jeanphilippeD**
To allow tests that load Parsec using `Parsec::from_parsed_contents`
when a DKG is in progress, dump existing KeyGen as serialized data
and reload them.

This is a very basic facility, and we may want to update it in the
future to provide more useful information in the dump-graphs.

Update dot_gen to support extracting a specific dump in a serie so
we can pick a graph that contain a DKG in progress.

Test:
Add new test that exercise dump and reload.

### CODE REVIEW

File: dot_gen/src/main.rs [hunk 1]
```diff
@@ -545,7 +559,7 @@ struct Scenario {
     name: String,
     seed: RngChoice,
     schedule_fn: Box<FnMut(&mut Environment) -> Schedule>,
-    files: BTreeMap<String, String>,
+    files: BTreeMap<(String, Option<u32>), String>,
```
Comment (fizyk20) [inline dot_gen/src/main.rs]: The optional `u32` could use a comment as to what is its purpose - maybe even replacing this tuple with a proper struct could be a good idea.

File: src/parsec.rs [hunk 2]
```diff
@@ -2360,6 +2362,19 @@ impl Parsec<Transaction, PeerId> {
                 .map(|info| info.consensused = true);
         }
 
+        if let Some(serialized_key_gens_and_next_id) =
+            &parsed_contents.serialized_key_gens_and_next_id
+        {
+            // `deserialise` require `Serialize` which is only available with `dump-graphs`
```
Comment (fizyk20) [inline src/parsec.rs]: Maybe you could elaborate on why we need to deserialise things outside of the `dump-graphs` functionality (and also maybe how `Cursor` helps here).

Comment (jeanphilippeD) [reply src/parsec.rs]: I'll try to clarify, in comment. Cursor is really just a way to get std::io::Read object so we can use that specific function from maidsafe_utils::serialisation. The problem with the one that take a vector is that it is over constrained and require Serialize.

We need to deserialize everytime we are loading dot file as part of `testing` feature which does not require dump-graphs.

File: dot_gen/src/main.rs (hunk 1 above)
Comment (jeanphilippeD) [reply dot_gen/src/main.rs]: Turn out there is an existing trick that I can use here, so I will not add that functionality:
see functional_tests::unpolled_and_unconsensused_observations
using "Alice-005" as the peer name.

File: src/dev_utils/dot_parser.rs [hunk 3]
```diff
@@ -127,18 +128,29 @@ fn parse_file(ctx: &Rc<ParserCtx>) -> Parser<u8, ParsedFile> {
         + parse_peer_list(ctx)
         + parse_consensus_mode()
         + parse_secure_rng()
+        + parse_key_gens()
         + parse_graph()
         + parse_meta_election(ctx)
         - parse_end())
     .map(
-        |(((((our_id, peer_list), consensus_mode), secure_rng_values), graph), meta_election)| {
+        |(
```
[the remark this answers is not in the document]
Comment (jeanphilippeD) [reply src/dev_utils/dot_parser.rs]: Yes. Well it is a bit messy, I wonder if there is a better way to express that without writting down that scary tuple of tuple.

[the remark this answers is not in the document]
Comment (maqi) [reply src/dev_utils/dot_parser.rs]: maybe another good reason to use a proper struct  :)

[the remark this answers is not in the document]
Comment (jeanphilippeD) [reply src/dev_utils/dot_parser.rs]: Not really an option here as it is what we get from the parsing infrastructure, but I'll see if I can improve how we destructure it now that it has become a very big tuple.

**Comment by jeanphilippeD** [comment]
Fixed review comments.

**Comment by jeanphilippeD** [comment]
Fixed comments, It took me 2 tries to fix it and It only show the last force push now for reasons.
So I forced push the original again, and force push the new version again so it shows all that has changed.

File: src/parsec.rs [hunk 4]
```diff
@@ -2360,6 +2362,22 @@ impl Parsec<Transaction, PeerId> {
                 .map(|info| info.consensused = true);
         }
 
+        if let Some(serialized_key_gens_and_next_id) =
+            &parsed_contents.serialized_key_gens_and_next_id
+        {
+            // `serialisation::deserialise` is over constrained and require `Serialize` which is
+            // only available with `dump-graphs` enable (as it is not safe to dump secrets).
```
Comment (fizyk20) [inline src/parsec.rs]: Typo, "enable" → "enabled" ;)

**Comment by jeanphilippeD** [comment]
Fixed typo.

**Pull request merged**